Race-condition in templating

While html/template `Execute` is concurrency-safe, `tpl.Funcs` isn't, meaning that concurrent requests might corrupt the funcMap, resulting in wrong translations at best, or a panic at worst. This can be observed by running the following test:

engine.go:90-114 calls tpl.Funcs(...) on the shared *template.Template on every request to bind the per-request printer (t, plural, elapsed). html/template Execute is concurrency-safe, but Funcs() concurrent with Execute() is not — under concurrent requests this is a data race that can corrupt the funcMap or panic.

```go
package template // import "miniflux.app/v2/internal/template"

import (
	"bytes"
	"sync"
	"testing"
)

func TestRenderConcurrency(t *testing.T) {
	engine := NewEngine("")
	engine.ParseTemplates()

	const templateName = "offline.html"
	languages := []string{"en_US", "fr_FR", "de_DE", "es_ES"}

	// Compute the expected output of each language sequentially first, before
	// any concurrency, so the comparison baseline is race-free.
	expected := make(map[string][]byte, len(languages))
	for _, language := range languages {
		expected[language] = engine.Render(templateName, map[string]any{
			"language": language,
			"theme":    "light_serif",
		})
	}

	const goroutinesPerLanguage = 50

	var wg sync.WaitGroup
	for _, language := range languages {
		for range goroutinesPerLanguage {
			wg.Add(1)
			go func(language string) {
				defer wg.Done()
				got := engine.Render(templateName, map[string]any{
					"language": language,
					"theme":    "light_serif",
				})
				if !bytes.Equal(got, expected[language]) {
					t.Errorf("Render(%q) for language %q produced output that does not match the sequential baseline", templateName, language)
				}
			}(language)
		}
	}
	wg.Wait()
}
```

```console
$ go test miniflux.app/v2/internal/template -run TestRenderConcurrency
--- FAIL: TestRenderConcurrency (0.03s)
    engine_test.go:51: Render("offline.html") for language "en_US" produced output that does not match the sequential baseline
    engine_test.go:51: Render("offline.html") for language "en_US" produced output that does not match the sequential baseline
    engine_test.go:51: Render("offline.html") for language "en_US" produced output that does not match the sequential baseline
    engine_test.go:51: Render("offline.html") for language "es_ES" produced output that does not match the sequential baseline
    engine_test.go:51: Render("offline.html") for language "en_US" produced output that does not match the sequential baseline
    engine_test.go:51: Render("offline.html") for language "en_US" produced output that does not match the sequential baseline
    engine_test.go:51: Render("offline.html") for language "fr_FR" produced output that does not match the sequential baseline
    engine_test.go:51: Render("offline.html") for language "en_US" produced output that does not match the sequential baseline
FAIL
FAIL	miniflux.app/v2/internal/template	0.036s
FAIL
$
```

The fix is to stop mutating shared state per request. The obvious way to do this would be to call `tpl.Clone()`, but this isn't free performance-wise, albeit I didn't benchmark it, or to pass the printer through the template data and make t/plural/elapsed read it.

--- a/internal/template/engine.go
+++ b/internal/template/engine.go
@@ -94,6 +94,13 @@ func (e *Engine) Render(name string, data map[string]any) []byte {
 		panic("The template " + name + " does not exists.")
 	}
 
+	// Clone the template so the per-request, language-specific functions below
+	// are bound on a private copy. The shared template stored in e.templates is
+	// only ever cloned (never executed directly), so concurrent requests no
+	// longer race on its function map, which previously could cause a response
+	// to be rendered with another concurrent request's language.
+	tpl = template.Must(tpl.Clone())
+
 	printer := locale.NewPrinter(data["language"].(string))
 
 	// Functions that need to be declared at runtime.

--- a/internal/template/engine_test.go
+++ b/internal/template/engine_test.go
@@ -1,0 +1,62 @@
+// SPDX-FileCopyrightText: Copyright The Miniflux Authors. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package template // import "miniflux.app/v2/internal/template"
+
+import (
+	"bytes"
+	"sync"
+	"testing"
+)
+
+// TestRenderConcurrency renders the same template concurrently in different
+// languages. Because Render binds per-request, language-specific functions
+// ("t", "plural", "elapsed") onto the template, doing so on a shared template
+// while other requests execute it corrupts the output: a request can be served
+// another request's language. Each concurrent render must match the output of
+// the equivalent sequential render for its language.
+func TestRenderConcurrency(t *testing.T) {
+	engine := NewEngine("")
+	engine.ParseTemplates()
+
+	languages := []string{"en_US", "fr_FR", "de_DE", "es_ES", "pt_BR", "ru_RU", "zh_CN", "it_IT"}
+
+	newData := func(language string) map[string]any {
+		return map[string]any{"language": language, "theme": "system_serif"}
+	}
+
+	// Establish the expected output for each language sequentially.
+	expected := make(map[string][]byte, len(languages))
+	for _, language := range languages {
+		expected[language] = engine.Render("offline.html", newData(language))
+	}
+
+	const iterations = 300
+
+	var wg sync.WaitGroup
+	var mu sync.Mutex
+	mismatches := make(map[string]int)
+
+	for i := 0; i < iterations; i++ {
+		language := languages[i%len(languages)]
+		wg.Add(1)
+		go func(language string) {
+			defer wg.Done()
+			got := engine.Render("offline.html", newData(language))
+			if !bytes.Equal(got, expected[language]) {
+				mu.Lock()
+				mismatches[language]++
+				mu.Unlock()
+			}
+		}(language)
+	}
+	wg.Wait()
+
+	if len(mismatches) > 0 {
+		total := 0
+		for _, n := range mismatches {
+			total += n
+		}
+		t.Fatalf("concurrent Render produced wrong output for %d/%d requests (wrong-language translations); per-language mismatches: %v", total, iterations, mismatches)
+	}
+}

